File: ios/Classes/lib/awesome_notifications_core/utils/CronUtils.swift
//
//  CronUtils.swift
//  awesome_notifications
//
//  Created by Rafael Setragni on 22/09/20.
//

import Foundation

public final class CronUtils {

    public var fixedNowDate:Date?

    public let validDateFormat:String = Definitions.DATE_FORMAT

    public func getInitialCalendar() -> Calendar {
        var calendar:Calendar = Calendar.current
        calendar.timeZone = DateUtils.utcTimeZone
        return calendar
    }
    
    /// https://www.baeldung.com/cron-expressions
    /// <second> <minute> <hour> <day-of-month> <month> <day-of-week> <year>
    public func getNextCalendar(
        initialDateTime:String?,
        crontabRule:String?
    ) -> Date? {

        if(StringUtils.isNullOrEmpty(initialDateTime) && StringUtils.isNullOrEmpty(crontabRule))
        { return nil }

        var now:Date, delayedNow:Date
        if(fixedNowDate == nil){
            now = DateUtils.getUTCDateTime()
        } else {
            now = fixedNowDate!
        }

        var initialScheduleDay:Date?
        if (initialDateTime == nil) {
            initialScheduleDay = now
        }
        else {
            initialScheduleDay = DateUtils.stringToDate(initialDateTime, timeZone: DateUtils.localTimeZone.identifier)
        }

        // if initial date is a future one, show in future. Otherwise, show now
        switch (now.compare(initialScheduleDay!)) {

            case .orderedDescending: // if initial date is not a repetition and is in the past, do not show
                if(StringUtils.isNullOrEmpty(crontabRule))
                { return nil }
                break

            case .orderedSame: // if initial date is right now, shows now
                break
            
            case .orderedAscending: // if initial date is in future, shows in future
                if(StringUtils.isNullOrEmpty(crontabRule))
                { return initialScheduleDay }
                break
            
            default:
                break
        }

        delayedNow = applyToleranceDate(now);

        if (!StringUtils.isNullOrEmpty(crontabRule)) {

            if(CronExpression.validate(cronExpression: crontabRule!)) {
                
                do {
                    let cronExpression:CronExpression = try CronExpression.init(crontabRule!)
                    let nextSchedule:Date? = cronExpression.getNextValidDate(referenceDate: now)

                    if (nextSchedule != nil && delayedNow.compare(nextSchedule!) == ComparisonResult.orderedAscending) {
                        return nextSchedule
                    } else {
                        // if there is no more valid dates, remove the repetitions
                        return nil
                    }

                } catch {
                    
                }
            }

            return nil
        }

        return delayedNow
    }

    /// Processing time tolerance
    public func applyToleranceDate(_ initialScheduleDay: Date) -> Date {
        let delayedScheduleDay:Date = initialScheduleDay.addingTimeInterval(TimeInterval(1))
        return delayedScheduleDay
    }
}

File: example/android/android_core/src/main/java/me/carda/awesome_notifications/core/utils/CronUtils.java
package me.carda.awesome_notifications.core.utils;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import java.text.ParseException;
import java.util.Calendar;
import java.util.Date;
import java.util.TimeZone;

import me.carda.awesome_notifications.core.exceptions.AwesomeNotificationsException;
import me.carda.awesome_notifications.core.externalLibs.CronExpression;

public final class CronUtils {

    public static String validDateFormat = "yyyy-MM-dd HH:mm:ss";

    /// https://www.baeldung.com/cron-expressions
    /// <second> <minute> <hour> <day-of-month> <month> <day-of-week> <year>
    @Nullable
    public static Calendar getNextCalendar(
        @NonNull Calendar fixedNowDate,
        @NonNull String crontabRule,
        @NonNull TimeZone timeZone
    ) throws AwesomeNotificationsException {

        if(
            fixedNowDate == null ||
            timeZone == null ||
            StringUtils.getInstance().isNullOrEmpty(crontabRule)
        ) return null;

        if(CronExpression.isValidExpression(crontabRule)) {
            try {
                CronExpression cronExpression = new CronExpression(crontabRule);
                cronExpression.setTimeZone(timeZone);
                Date nextSchedule =
                        cronExpression
                                .getNextValidTimeAfter(fixedNowDate.getTime());

                Calendar delayedNow = applyToleranceDate(fixedNowDate);
                if (nextSchedule != null && nextSchedule.compareTo(delayedNow.getTime()) >= 0) {

                    Calendar calendar = Calendar.getInstance();
                    calendar.setTimeZone(timeZone);
                    calendar.setTime(nextSchedule);
                    return calendar;

                } else {
                    // if there is no more valid dates, remove the repetitions
                    return null;
                }

            } catch (ParseException e) {
                e.printStackTrace();
            }
        }

        return null;
    }

    /// Processing time tolerance
    public static Calendar applyToleranceDate(Calendar initialScheduleDay) {
        Calendar shifted = (Calendar) initialScheduleDay.clone();
        shifted.set(Calendar.MILLISECOND, 0);
        return shifted;
    }
}

